### PR TITLE
ttt with a cli

### PR DESCRIPTION
**Pull request by liza-pizza**
Added Cli to tic tac toe.

### CODE REVIEW

**Comment by nid90** [comment]
A general observation: add a bit more words to the commit messages. it is not just about adding or updating files or fns, it is about giving a brief but thorough description of the changes going into that commit.

ex: `add console` is quite un-descriptive,  I would reword that as `add a console namespace to tic tac toe to add a CLI to the game`, or something to that effect.

this has some good ideas: https://initialcommit.com/blog/git-commit-messages-best-practices

File: practice/ttt/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Tic Tac Toe
```
Comment (kirang89) [inline practice/ttt/README.md:1]: suggestion: Try to think of the README as the first thing a potential user might be looking at. What would that person want to see here?

Some potential questions:
- What does this repo do?
- How do I set this up in my machine?
- How do I test this out?
- How to I play the game?

Thinking from this perspective will help pen down a more thoughtful README.

File: practice/ttt/board.clj [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  (keys position-to-coordinate))
+
+
+(def positions-string
+  (mo/matrix->string (partition size (sort valid-positions))))
+
+
+(def valid-game-pieces-set #{:o :x :e})
+
+(def game-pieces-set #{:o :x})
```
Comment (kirang89) [inline practice/ttt/board.clj]: question: Should we consider `:x` and `:o` as players rather than pieces? Even the [wiki](https://en.wikipedia.org/wiki/Tic-tac-toe) thinks of it that way. What do you think?

File: practice/ttt/board.clj [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+       (clojure.set/subset? (set (flatten board)) valid-game-pieces-set)))
+
+
+(defn update
+  [board coordinate game-piece]
+  (when (valid? board)
+    (assoc-in board coordinate game-piece)))
+
+
+(defn winner-of-collection
```
Comment (kirang89) [inline practice/ttt/board.clj:58]: question: What does `collection` refer to?

File: practice/ttt/user_input.clj [hunk 4]
```diff
@@ -0,0 +1,30 @@
+(ns ttt.user-input
```
Comment (kirang89) [inline practice/ttt/user_input.clj:1]: suggestion/question: What do you think about naming this layer `cli` instead of `user-input`? The functions here are very specific to parsing user input _from the command line_. We might have an entirely different set of validations for user input if we were to create a web UI interface for the game, and so this namespace might not entirely make sense at that point. It's a stretch of course (since we don't plan on building a web UI), but still something to think about.

File: practice/ttt/matrix_operations.clj [hunk 5]
```diff
@@ -47,3 +47,11 @@
     (transpose-matrix matrix)
     (get-diagonals-of-matrix matrix)))
 
+
+(defn matrix->string
+  [matrix]
+  (apply str (for [row matrix]
+               (str
+                 (apply str "\n"
+                        (interpose " " row))))))
```
Comment (kirang89) [inline practice/ttt/matrix_operations.clj]: nitpick: Using a thread last macro might make the section more readable:
```clojure
(->>  row
     (interpose " ")
     (apply str "\n")
     str)))
```
Also, do we need to `str` at the end if we're doing `apply str` in the previous step?

File: practice/ttt/board.clj (hunk 3 above)
Comment (liza-pizza) [reply practice/ttt/board.clj:58]: the function can find the winner of any collection as opposed to a 3x3 board. Used the word collection to generalize

File: practice/ttt/user_input.clj (hunk 4 above)
Comment (liza-pizza) [reply practice/ttt/user_input.clj:1]: I understand, the user-input namespace is never reading directly from the cli. This particular namespace has no idea that a cli exists, it's just parsing input that is not necessarily from the command line. Thats why I think naming it user-input works better.
